disk-encryption-tool

```diff
@@ -221,9 +233,67 @@ make_ro()
 	btrfs prop set -t s "$mp" ro true
 }
 
+enroll_fido2()
+{
+	if [ ! -e /usr/bin/systemd-cryptenroll ]; then
+		log_info "missing systemd-cryptenroll in initrd"
+		return
+	fi
+
+	devices="$(systemd-cryptenroll --fido2-device=list 2>&1)"
+	if grep -q "No FIDO2" <<<"$devices"; then
+		echo "No FIDO2 devices found"
+		return
+	fi
+
+	echo -e "\e[1mFIDO2 device present. If registered will take precedence from TPM2\e[m"
+
+	umask 066; echo -n "$password" > "$tmpdir"/pw; umask 022
+	systemd-cryptenroll --unlock-key-file="$tmpdir"/pw --fido2-device=auto "$blkpart" && with_fido2=1
+	rm "$tmpdir"/pw
+}
+
+enroll_tpm2()
+{
+	# TODO: very suboptimal approach as it requires the creation
+	# of a new initrd.  The initrd should not be modified, so it
+	# can be measured via PCR9.  The JSON file should then be
+	# stored in the ESP, close to the boot loader entry
```

Comment:
Does this actually have to happen in initrd? Looks like it could be done in jeos-firstboot too?

Reply:
It need to happen in initrd, as it is the one that will unlock the sysroot. The thing that is not required is that it needs to re-do initrd to put the key and json in the initrd, as a dracut module can read it from the ESP, and give it into the systemd-cryptsetup later